### DISCUSSION TITLE
How to use MinimumEnclosingCircle and PythonTransform?

### DISCUSSION THREAD

**Discussion by mmower2000** [bonsai-rx/bonsai#1704] [Q&A] [answered]
Hi there, I'd appreciate a little help with a couple nodes I can't quite work out how to use properly in bonsai! I have a physical disk which will be turning with three equally spaced LEDs attached and I want to track the LEDs with video in order to get an estimation on the orientation of the disk.
 
1. How do you use 'MinimumEnclosingCircle'? In order to estimate the centre point of the disk (to later use in calculating the angle of LED movement) I want to estimate the circle size by using the three LEDs. I'd then use this to find its centroid. I am currently trying to give it three xy coordinates from the tracked LEDs, but it says 'no method overload found for given arguments' and I cannot seem to find what inputs do actually work with this node.

2. How do I use the' PythonTransform'? I want to use the python transform to calculate the changing angles of the LEDs compared to their starting point, to find the orientation change. However, I cannot seem to find out what inputs and return values will work. Ideally I want to enter in tuples of xy coordinates (centre point, LED starting, LED current) and have a float output for the angle.
 
I hope this made sense. Thanks for your time!

**Comment by bruno-f-cruz** [comment]
Please add the workflow and any data that may be useful to help you! Thanks

**Reply by mmower2000** [reply]
So here is the workflow and an example video. The undistort and affline transform are not in use yet as the camera isn't fitted properly yet fyi...just trying to get the pipeline ready. The three steams of processing is for each LED. I then want to find the centre and use it in python to work out the orientation of the circle. For now I will probably use the starting frame as the original LED position for each LED, I will work out the angle change for each LED with this method and then average this to get orientation. Thanks again!

https://github.com/bonsai-rx/bonsai/assets/122468390/36361b99-70b8-45aa-bfb4-3f8a56d157dc


![LED_tracking](https://github.com/bonsai-rx/bonsai/assets/122468390/f0023378-48fe-4950-8041-460c3696fac0)
[TestColourTracking.csv](https://github.com/bonsai-rx/bonsai/files/14463351/TestColourTracking.csv)

**Reply by bruno-f-cruz** [reply]
Please add the workflow file as asked in #864. Thanks!

**Reply by mmower2000** [reply]
[LightTracking.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/14463585/LightTracking.bonsai.zip)
Okay here we go, apologies for the confusion

**Reply by bruno-f-cruz** [reply, marked as answer]
Checking the documentation of `MininumEnclosingCircle` shows that you probably want to use the overload with an Array of points as input (https://bonsai-rx.org/docs/api/Bonsai.Vision.MinimumEnclosingCircle.html#Bonsai_Vision_MinimumEnclosingCircle_Process_System_IObservable_OpenCV_Net_Point2f____)

You can modify your workflow to create a Point2f[3] array for each frame and feed it to the operator. See the solution attached.

Regarding the PythonTransform, I am not sure what you are trying to do, but I edited your example to show how to return a Tuple[float, float, float] and how to access the incoming tuple elements.
[LightTracking.zip](https://github.com/bonsai-rx/bonsai/files/14463687/LightTracking.zip)

Hope it helps!

**Reply by mmower2000** [reply]
Okay that's working great, thanks a lot for the quick reply!  I am very new to Bonsai so forgive me if this is a really basic question, but how did you get MyTuple(both types) and the CreateArray I can't seem to find these nodes in the Toolbox or as output/Externalize property. I can see that the green/source MyTuple has become a subject in your corrected version. Thanks!

**Reply by bruno-f-cruz** [reply]
Here are the respective docs:
[`CreateArray` is a `SelectMany` operator](https://bonsai-rx.org/docs/articles/reactive-selectmany.html?q=selectmany)

[`MyTuple` is a `Subject`](https://bonsai-rx.org/docs/articles/subjects.html)

**Reply by mmower2000** [reply]
Ah of course I see this now. Thanks Again Bruno!
